filebeat/input/filestream/internal/input-logfile/harvester.go

```diff
@@ -60,11 +64,20 @@ type defaultHarvesterGroup struct {
 // Run starts the Harvester for a Source.
 func (hg *defaultHarvesterGroup) Run(ctx input.Context, s Source) error {
 	log := ctx.Logger.With("source", s.Name())
-	log.Debug("Starting harvester for file")
 
 	harvesterCtx, cancelHarvester := context.WithCancel(ctxtool.FromCanceller(ctx.Cancelation))
 	ctx.Cancelation = harvesterCtx
 
+	// this blocks Run until a new reader can be started for the file
+	limitErr := hg.limitSem.Acquire(harvesterCtx, 1)
```

Comment:
The semaphore is not released correctly all the time, there are other steps than can fail. A more aggressive pattern would be:

```
limitErr := hg.limitSem.Acquire(harvesterCtx, 1)
...

setupErr := func() (err error) {
  defer func() {
    if err != nil {
      cancelHarvester()
      hg.limitSem.Release(1)
    }
  }()

  // all setup code
}()
if setupErr != nil {
 ...
}

go func() {
}()
```

moving the 'setup' into its own body and capturing the error in `defer` allows you to always run the code in defer no matter at which line an error occured (eases cleanup in case developers add more sites that can fail during setup)